### PR TITLE
feat: time it custom callback 

### PR DESCRIPTION
**Pull request by AlexXi19**
## Description 
Custom callback function v1: allow users to define a custom callback function. This is v1 because the callbacks are executed in the normal code path, which can make the application slower and affect the collected latencies in the case of a nested usage. Might be better to change this to an event/subscriber based design. 

## Changes 
- Create new builder crate to init the callback function 
- Change the macro crate to use the initialized config from the builder crate 
- Run cargo +nightly fmt 

## Performance  

Comparing the performance of using a read lock vs. lazy init, the behavior is consistent with the expected behavior of lazy init: higher up front cost (though haven't investigated too deeply into where the cost is coming from, presumable acquiring the read lock + whatever the lazy init crate is doing), but lower latencies for subsequent usage. 

(Collected in debug mode)

Read lock: 

```
Lock Duration: 58.334µs
[Custom Message] Time Elapsed: 505ms
I still have ownership of Defined in macro!!!
Lock Duration: 30.792µs
[Custom Message] Time Elapsed: 505ms
I still have ownership of Defined in macro!!!
Lock Duration: 46.5µs
[Custom Message] Time Elapsed: 505ms
I still have ownership of Defined in macro!!!
```

Lazy init 

```
Config Duration: 138.416µs
[Custom Message] Time Elapsed: 505ms
I still have ownership of Defined in macro!!!
Config Duration: 5.125µs (wow so fast)
[Custom Message] Time Elapsed: 503ms
I still have ownership of Defined in macro!!!
Config Duration: 4.75µs
[Custom Message] Time Elapsed: 503ms
I still have ownership of Defined in macro!!!
Config Duration: 31.042µs
[Custom Message] Time Elapsed: 506ms
I still have ownership of Defined in macro!!!
Config Duration: 4.375µs
[Custom Message] Time Elapsed: 502ms
I still have ownership of Defined in macro!!!
```

### CODE REVIEW

File: time-it-builder/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+name = "time-it-builder"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+lazy_static = "1.4.0"
+lazy-init = "0.5.1"
+parking_lot = "0.12.1"
```
Comment (tStreichenberger) [inline time-it-builder/Cargo.toml:11]: this seems huge. Could be good as an addition to rust hsc in places where we are forced to use sync locks, which we actually have in a few places in lua

File: time-it-builder/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+use std::{sync::Arc, time::Duration};
+
+lazy_static::lazy_static! {
+    static ref TIMER_CONFIG: RwLock<TimeItConfig> = RwLock::new(TimeItConfig::default());
+    static ref LAZY_TIMER_CONFIG: Lazy<Arc<TimeItConfig>> = Lazy::new();
+}
+
+#[derive(Clone)]
+pub struct TimeItConfig {
+    pub action: Arc<Box<dyn Fn(Duration) -> () + Send + Sync>>,
```
Comment (tStreichenberger) [inline time-it-builder/src/lib.rs]: do you need an Arc and box? can probably just use arc no?

File: time-it-builder/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+pub struct TimeItBuilder {}
+
+impl TimeItBuilder {
+    pub fn new() -> Self {
+        TimeItBuilder {}
+    }
+
+    /// This function is called when the timer is done.
+    /// The first argument is the name of the timer, the second is the duration.
+    pub fn time_it(self, action: Box<dyn Fn(Duration) -> () + Send + Sync>) -> Self {
```
Comment (tStreichenberger) [inline time-it-builder/src/lib.rs]:
```suggestion
    pub fn time_it(&self, action: Box<dyn Fn(Duration) -> () + Send + Sync>) -> &Self {
```
should make these references. I made the mistake of not doing refernces on a builder in the rust hsc and it lead to lines of code like this some times, if a user wanted to assign just one value, for example in a conditional
```
let builder = builder.something("hi");
```

File: time-it-macros/src/lib.rs [hunk 4]
```diff
@@ -19,13 +16,14 @@ pub fn time_it(to_time: TokenStream) -> TokenStream {
     quote! {
         let #start_ident = std::time::Instant::now();
         #to_time
-        println!("Operation took {}ms", #start_ident.elapsed().as_millis());
+        let duration = #start_ident.elapsed();
+        let config = time_it::get_config();
+        let action = config.action.clone();
+        action(duration);
```
Comment (tStreichenberger) [inline time-it-macros/src/lib.rs]: nit: could shorten this to `time_it::action(duration)` which just does these 3 lines under the hood

**Review by tStreichenberger — APPROVED** [review]

File: time-it-macros/src/lib.rs (hunk 4 above)
Comment (tStreichenberger) [reply time-it-macros/src/lib.rs]: would also remove need for arc clone because the static action function could get config directly and just use ref

Comment (AlexXi19) [reply time-it-macros/src/lib.rs]: Added the action method for a better interface but observed no change in latency. Will keep the Arc here just in case we are adding more fields into the config and want to pass it out.
